src/integration/java/io/pinecone/integration/dataPlane/UpsertAndQueryServerlessTest.java

```diff
@@ -0,0 +1,238 @@
+package io.pinecone.integration.dataPlane;
+
+import com.google.protobuf.Struct;
+import io.pinecone.clients.AsyncIndex;
+import io.pinecone.clients.Index;
+import io.pinecone.clients.Pinecone;
+import io.pinecone.exceptions.PineconeValidationException;
+import io.pinecone.helpers.RandomStringBuilder;
+import io.pinecone.proto.DescribeIndexStatsResponse;
+import io.pinecone.unsigned_indices_model.QueryResponseWithUnsignedIndices;
+import io.pinecone.unsigned_indices_model.ScoredVectorWithUnsignedIndices;
+import org.junit.jupiter.api.AfterAll;
+import org.junit.jupiter.api.BeforeAll;
+import org.junit.jupiter.api.Test;
+import org.openapitools.client.model.IndexModelSpec;
+
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.List;
+import java.util.concurrent.ExecutionException;
+
+import static io.pinecone.helpers.AssertRetry.assertWithRetry;
+import static io.pinecone.helpers.BuildUpsertRequest.*;
+import static io.pinecone.helpers.IndexManager.*;
+import static org.junit.jupiter.api.Assertions.assertEquals;
+
+public class UpsertAndQueryServerlessTest {
+    private static Index index;
+    private static AsyncIndex asyncIndex;
+    private static final int dimension = 3;
+    private static final Struct emptyFilterStruct = Struct.newBuilder().build();
+
+    @BeforeAll
+    public static void setUp() throws IOException, InterruptedException {
+        String apiKey = System.getenv("PINECONE_API_KEY");
+        String indexType = IndexModelSpec.SERIALIZED_NAME_SERVERLESS;
+        Pinecone pinecone = new Pinecone(apiKey);
+
+        String indexName = findIndexWithDimensionAndType(pinecone, dimension, indexType);
+        if (indexName.isEmpty()) indexName = createNewIndex(pinecone, dimension, indexType);
```

Comment:
This is the first serverless data plane test, and this is searching for any index of `dimension=3` with a serverless spec, and creating one if it doesn't exist. Is that right? I was mainly trying to think if this could conflict with any other serverless indexes with the same `dimension`, but I think this should be ok for now.

Reply:
That is correct and since all of the data plane tests are of dimension 3, it shouldn't have any conflict with control plane.